pkg/streamingpromql/operators/functions/rate_increase.go

```diff
@@ -289,3 +280,94 @@ func rateSeriesValidator() RangeVectorSeriesValidationFunction {
 		lastCheckedMetricName = metricName
 	}
 }
+
+func delta() RangeVectorStepFunction {
+	return func(step *types.RangeVectorStepData, rangeSeconds float64, emitAnnotation types.EmitAnnotationFunc) (float64, bool, *histogram.FloatHistogram, error) {
+		fHead, fTail := step.Floats.UnsafePoints()
+		fCount := len(fHead) + len(fTail)
+
+		hHead, hTail := step.Histograms.UnsafePoints()
+		hCount := len(hHead) + len(hTail)
+
+		if fCount > 0 && hCount > 0 {
+			// We need either at least two histograms and no floats, or at least two floats and no histograms to calculate a rate.
+			// Otherwise, emit a warning and drop this sample.
+			emitAnnotation(annotations.NewMixedFloatsHistogramsWarning)
+			return 0, false, nil, nil
+		}
+
+		if fCount >= 2 {
+			// TODO: just pass step here? (and below)
+			val := floatDelta(fCount, fHead, fTail, step.RangeStart, step.RangeEnd, rangeSeconds)
+			return val, true, nil, nil
+		}
+
+		if hCount >= 2 {
+			val, err := histogramDelta(hCount, hHead, hTail, step.RangeStart, step.RangeEnd, rangeSeconds, emitAnnotation)
+			if err != nil {
+				err = NativeHistogramErrorToAnnotation(err, emitAnnotation)
+				return 0, false, nil, err
+			}
+			return 0, false, val, nil
+		}
+
+		return 0, false, nil, nil
+	}
+}
+
+func floatDelta(fCount int, fHead []promql.FPoint, fTail []promql.FPoint, rangeStart int64, rangeEnd int64, rangeSeconds float64) float64 {
+	firstPoint := fHead[0]
+	fHead = fHead[1:]
+
+	var lastPoint promql.FPoint
+	if len(fTail) > 0 {
+		lastPoint = fTail[len(fTail)-1]
+	} else {
+		lastPoint = fHead[len(fHead)-1]
+	}
+
+	delta := lastPoint.F - firstPoint.F
+	return calculateFloatRate(false, false, rangeStart, rangeEnd, rangeSeconds, firstPoint, lastPoint, delta, fCount)
+}
+
+func histogramDelta(hCount int, hHead []promql.HPoint, hTail []promql.HPoint, rangeStart int64, rangeEnd int64, rangeSeconds float64, emitAnnotation types.EmitAnnotationFunc) (*histogram.FloatHistogram, error) {
+	firstPoint := hHead[0]
+	hHead = hHead[1:]
+
+	var lastPoint promql.HPoint
+	if len(hTail) > 0 {
+		lastPoint = hTail[len(hTail)-1]
+	} else {
+		lastPoint = hHead[len(hHead)-1]
+	}
+
+	if firstPoint.H.CounterResetHint == histogram.GaugeType || lastPoint.H.CounterResetHint == histogram.GaugeType {
+		emitAnnotation(annotations.NewNativeHistogramNotCounterWarning)
+	}
+
+	desiredSchema := firstPoint.H.Schema
+	if lastPoint.H.Schema < desiredSchema {
+		desiredSchema = lastPoint.H.Schema
+	}
+
+	usingCustomBuckets := firstPoint.H.UsesCustomBuckets()
+	if lastPoint.H.UsesCustomBuckets() != usingCustomBuckets {
```

Comment:
[nit] This could be simplified to:

```suggestion
	if firstPoint.H.UsesCustomBuckets() != lastPoint.H.UsesCustomBuckets() {
```